core/c4diagram.go

```diff
@@ -0,0 +1,355 @@
+	switch e {
+	case 0:
+		return '-'
+	case 1:
+		return '_'
+	default:
+		return '?'
+	}
+}
+
+func trimQuotes(s string) string {
+	if s[:1] == `"` {
+		s = s[1:]
+	}
+	if s[len(s)-1:] == `"` {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+func diagramFooter2UML(graph DiagramGraph) string {
+	if graph.Footer == "" {
+		return `footer "generated by diagramastext.dev - %date('yyyy-MM-dd')"`
+	}
+	return `footer "` + stringCleaner(graph.Footer) + `"`
+}
+
+func diagramNode2UML(n *Node) (string, error) {
+	if n.ID == "" {
+		return "", errors.New("node must be identified: 'id' attribute")
+	}
+
+	o := "Container"
+
+	switch n.IsQueue && n.IsDatabase {
+	case true:
+	case false:
+		if n.IsQueue {
+			o += "Queue"
+		}
+
+		if n.IsDatabase {
+			o += "Db"
+		}
+	}
+
+	if n.External {
+		o += "_Ext"
+	}
+
+	o += "(" + n.ID
+
+	label := n.Label
+	if label == "" {
+		label = n.ID
+	}
+	o += `, "` + stringCleaner(label) + `"`
+
+	if n.Technology != "" {
+		o += `, "` + stringCleaner(n.Technology) + `"`
+	}
+
+	o += ")"
+
+	return o, nil
+}
+
+func diagramLink2UML(l *Link) (string, error) {
+	if l.From == "" || l.To == "" {
+		return "", errors.New("link must specify the end nodes: 'from' and 'to' attributes")
+	}
+
+	o := "Rel"
+
+	if d := linkDirection(l.Direction); d != "" {
+		o += "_" + d
+	}
+
+	o += "(" + l.From + ", " + l.To
+
+	if l.Label != "" {
+		o += `, "` + stringCleaner(l.Label) + `"`
+	}
+
+	if l.Technology != "" {
+		o += `, "` + stringCleaner(l.Technology) + `"`
+	}
+
+	o += ")"
+
+	return o, nil
+}
+
+func linkDirection(s string) string {
+	switch s := strings.ToUpper(s); s {
+	case "LR":
+		return "R"
+	case "RL":
+		return "L"
+	case "TD":
+		return "D"
+	case "DT":
+		return "U"
+	default:
+		return ""
+	}
+}
+
+func stringCleaner(s string) string {
+	s = strings.TrimSpace(s)
+	s = trimQuotes(s)
+	s = strings.ReplaceAll(s, "\n", "\\n")
+	return s
+}
+
+// diagramGraph2plantUMLCode function to "transpile" the diagram definition graph to plantUML code as string.
+func diagramGraph2plantUMLCode(graph DiagramGraph) (string, error) {
+	if len(graph.Nodes) == 0 {
+		return "", errors.New("graph must contain at least one node")
+	}
+
+	o := `@startuml
+!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Container.puml`
+
+	o += "\n" + diagramFooter2UML(graph)
+
+	if graph.Title != "" {
+		o += "\n" + diagramTitle2UML(graph)
+	}
+
+	groups := map[string][]string{}
+	for _, n := range graph.Nodes {
+		containerStr, err := diagramNode2UML(n)
+		if err != nil {
+			return "", err
+		}
+
+		if _, ok := groups[n.Group]; !ok {
+			groups[n.Group] = []string{}
+		}
+		groups[n.Group] = append(groups[n.Group], containerStr)
+	}
+	o += diagramUMLSystemBoundary(groups)
+
+	for _, l := range graph.Links {
+		linkStr, err := diagramLink2UML(l)
+		if err != nil {
+			return "", err
+		}
+		o += "\n" + linkStr
+	}
+
+	o += "\n@enduml"
+
+	return o, nil
+}
+
+func diagramUMLSystemBoundary(v map[string][]string) string {
+	o := ""
```

Comment:
**suggestion** use `strings.Builder` to create this string efficiently.

Reply:
A good one for refactoring!

Reply:
**(nitpick)** when make suggestions, please share a code snippet, or a reference. Thanks!

Reply:
Added "FIXME" for later refactoring